Add google pay documentation (#69)
clearhaus/gateway-api-docs: source/index.html.md

not be included. 3-D Secure version 2 has some support for subsequent
authorizations.

An initial recurring authorization can also be made using the `applepay`,

> `googlepay` and `mobilepayonline` payment methods; subsequent recurring

Reviewer: ```suggestion
`googlepay` or `mobilepayonline` payment methods; subsequent recurring
```

````

Authorizations can be created using different payment methods:

> `card`, `applepay`, `googlepay`, `mobilepayonline`.

Reviewer: The ordering here aligns with the ordering below, but any reason for squeezing `googlepay` in (signaling some kind of ordering I don't understand) rather than adding it to the end?

Author: Just phonetically I liked the two `*pay` words to follow each other :)

  subsequent recurring authorization.
</p>

##### Method: `googlepay`

To accept a payment using Google Pay, the complete payment token, recipient ID
and derived shared secret, are required. Please refer to the [official
documentation][GooglePay-PaymentCryptography]. Only protocol version `ECv2` is
supported.

<dl class="dl-vertical">
  <dt>googlepay[token]
    <span class="type">[:json:]</span>
  </dt>
  <dd>
    Raw payment method token as received in response from Google. UTF-8 encoded
    serialization of a JSON dictionary.
  </dd>
  <dt>googlepay[shared_key]
    <span class="type">[:base64:]</span>
  </dt>
  <dd>
    The shared secret derived from the ephemeral public key and your private
    key.
  </dd>
  <dt>googlepay[recipient_id]

>     <span class="type">[\x20-\x7E]+

Reviewer: Since space is not allowed:
```suggestion
    <span class="type">[\x21-\x7E]+
```

But also, isn't there an upper limit on the length?

I went through the documentation I have at hand and couldn't find any.

>       <a target="_blank" href="http://en.wikipedia.org/wiki/ASCII#ASCII_printable_characters">ASCII printable characters excluding space</a>

Reviewer: ```suggestion
      <a target="_blank" href="http://en.wikipedia.org/wiki/ASCII#ASCII_printable_characters">ASCII printable characters</a> excluding space
```

🤷‍♂️

    </span>
  </dt>
  <dd>
    The Google Pay recipient of the payload, e.g. <code>merchant:0123456789</code>.
  </dd>
</dl>

<p class="alert alert-info">
  <b>Notice:</b> Signing is required to use the <code>googlepay</code> payment
  method.
  <br />
  <b>Notice:</b> An authorization made with <code>googlepay</code> may be fully
  3-D Secured, 3-D Secure attempted, or with no 3-D Secure; this is indicated by

>   the <code>eciIndicator</code> of the <code>paymentMethodDetails</code>.

Reviewer: According to https://developers.google.com/pay/api/web/guides/resources/payment-data-cryptography#cryptogram_3ds, `eciIndicator` (RAS Syndrome) is only present for Visa.

Author: I'm actually not sure that that is true, _but_, it may be true.

  <br />
  <b>Notice:</b> An authorization made with <code>googlepay</code> is

>   strongly authenticated (SCA in PSD2) if fully 3-D Secure.

Reviewer: I think we need to relax this statement. As a random example, Mastercard LI `kA` has status `Y` and is not SCA.

Author: I think you're better equipped to answer that than me :)

  subsequent recurring authorization.
</p>

##### Method: `googlepay`

To accept a payment using Google Pay, the complete payment token, recipient ID
and derived shared secret, are required. Please refer to the [official
documentation][GooglePay-PaymentCryptography]. Only protocol version `ECv2` is
supported.

<dl class="dl-vertical">
  <dt>googlepay[token]
    <span class="type">[:json:]</span>
  </dt>
  <dd>
    Raw payment method token as received in response from Google. UTF-8 encoded
    serialization of a JSON dictionary.
  </dd>
  <dt>googlepay[shared_key]
    <span class="type">[:base64:]</span>
  </dt>
  <dd>
    The shared secret derived from the ephemeral public key and your private
    key.
  </dd>
  <dt>googlepay[recipient_id]
    <span class="type">[\x21-\x7E]+
      <a target="_blank" href="http://en.wikipedia.org/wiki/ASCII#ASCII_printable_characters">ASCII printable characters</a> excluding space
    </span>
  </dt>
  <dd>
    The Google Pay recipient of the payload, e.g. <code>merchant:0123456789</code>.
  </dd>
</dl>

<p class="alert alert-info">
  <b>Notice:</b> Signing is required to use the <code>googlepay</code> payment
  method.
  <br />

>   <b>Notice:</b> An authorization made with <code>googlepay</code> is strongly
>   authenticated (SCA in PSD2) if <code>authMethod</code> is
>   <code>CRYPTOGRAM_3DS</code> and the Google Pay guidelines for
>   <i><a href="https://developers.google.com/pay/api/web/guides/resources/sca">sca</a></i>
>   have been followed.

Reviewer: Just a slight "link improvement":
```suggestion
  <b>Notice:</b> An authorization made with <code>googlepay</code> is strongly
  authenticated (SCA in PSD2) if <code>authMethod</code> is
  <code>CRYPTOGRAM_3DS</code> and
  <a href="https://developers.google.com/pay/api/web/guides/resources/sca">the
  Google Pay guidelines for SCA</a> have been followed.
```

What's the reason for the italic?

Author: The link doesn't look different from ordinary text, so I wanted to highlight it.

It's probably better to make it explicit with a link that looks like `[link]`. I'll do that.

>   If <code>authMethod</code> is <code>PAN_ONLY</code>, a 3-D Secure flow is
>   required for SCA.

Reviewer: ```suggestion
  If <code>authMethod</code> is <code>PAN_ONLY</code>, a 3-D Secure flow is
  necessary to obtain strong authentication.
```

Dunno which version is easier to understand correctly 🤷‍♂️

Author: I'd prefer to use "SCA", as it's probably easily recognizable as SCA related (and briefer).
But I have no strong opinion. You're welcome to merge if you prefer your edition.

  <br />
  <b>Notice:</b> An authorization made with <code>googlepay</code> cannot be a
  subsequent recurring authorization.
  <br />

>   <b>Notice:</b> The merchant ID for the <code>googlepay</code> test
>   environment is <code>12345678901234567890</code>.

Reviewer: ```suggestion
  <b>Notice:</b> The <code>recipient_id</code> for the <code>googlepay</code>
  test environment is <code>merchant:12345678901234567890</code>.
```

This is correct, right? If so, I think the added detail could perhaps be helpful 🙂

Author: Ah, good catch :+1: A remnant of an old version.
